### DISCUSSION TITLE
Using nanobind as Python bindings generator

### DISCUSSION THREAD

**Discussion by nicholasjng** [duckdb/duckdb#13008] [Ideas]
Hello! We (@maxmynter and I) are interested in developing (and eventually contributing) [nanobind](https://github.com/wjakob/nanobind/) Python bindings for duckdb if there is shared interest on your side.

The main benefits would be smaller binding sizes, less runtime overhead, and a builtin stub generator for binding extensions eliminating the need for mypy's stubgen. The new `nanobind::ndarray` submodule would also open up the possibility of streaming data into different types of arrays other than NumPy, like TF/PyTorch tensors or JAX arrays. See the [rationale](https://nanobind.readthedocs.io/en/latest/why.html) section in the docs for further information.

We have already spent a bit of time last week porting over some of the code in my fork (https://github.com/nicholasjng/duckdb/tree/nanobind-bindings) and checking in general how the current code would translate to nanobind. A few remarks and open questions my side:

* How would the pandas/numpy code look like with the `nb::ndarray` paradigm?
* I think all of the existing enum type casters can be removed by virtue of nanobind v2.0.0's builtin enum type casting functionality.
* Are there any objections to removing the current module locality of bindings? (That is unsupported in nanobind.)
* How do you feel about moving duckdb-python's packaging over to CMake also (maybe with `scikit-build-core` for PEP517 support)?

Looking forward to a discussion.
cc @Tishj

**Comment by Tishj** [comment]
Thanks for starting this work, I had a quick look through the branch and it seems to be a seamless transition 👍 

One concern I had however when I dug into nanobind is that they mention their ownership semantics are different from pybind.
Did you have a chance to look through these differences? I am worried this might result in hard-to-debug leaks/crashes - especially since we have our own wrappers of unique_ptr/shared_ptr and nanobind seems to be tied more closely to the `std` in that regard, or am I misreading that?

**Reply by nicholasjng** [reply]
That's also a bit of an unknown due to my limited experience with the C++ codebase - you also have a wrapper for Optional, which I'm guessing fills in for `std::optional`?

I think the main difference is that instead of the holder type / caster declarations, you return the shared pointer in your Python class' allocation method (i.e. `DuckDBPyConnection.__new__()`). This has implications on classes using `enable_shared_from_this` regarding reference counting, of which there are a few in duckDB IIRC, but I'm not sure what they are yet.

**Reply by Tishj** [reply]
No it doesn't stand in for `std::optional`, I understand why you'd think that, but it's merely cosmetic to be able to have `Optional<...>` in generated stubs.

**Reply by Tishj** [reply]
Exactly, there are implications because the models don't align, and without research into what those are, I don't feel comfortable making this switch.

To add on top of this, `enable_shared_from_this` is also custom because our `shared_ptr`+`weak_ptr` are custom.

**Comment by Tishj** [comment]
> How do you feel about moving duckdb-python's packaging over to CMake also (maybe with scikit-build-core for PEP517 support)?

I'm not super familiar with the internals of the python build, but it has been a painpoint for a while.
One notable problem is that in the `setup.py` we rely on python scripts that reside outside of the `tools/pythonpkg` folder, which does not play well with the whole "independent build" process where python will copy the entire project to a temporary directory to build

**Reply by nicholasjng** [reply]
I skimmed the `package_build.py`, and most of it is defining include and source path flags - should be doable in CMake. With the setuptools_scm version it's another thing, since the C++ file including the version seems to be autogenerated at build time, but that could also be done in `pyproject.toml` e.g. with a dynamic version field.

I am planning to attend the scikit-build-core dev monthly meeting this Friday and ask a bit about the build setup in CMake.

**Reply by Tishj** [reply]
We use CMake for every other facet of building, so I imagine there was a good reason this couldn't be used here. Whether that blocker still exists I don't know.
The code dates back to 4 years ago, there weren't really PRs back then so retracing those reasons is near impossible sadly.

**Comment by Tishj** [comment]
> Are there any objections to removing the current module locality of bindings?

I'm not really sure what that implies, I have seen the global/local nature of module definitions in pybind but never dug deeper into it

**Comment by nicholasjng** [comment]
Hey, just wanted to let you know that I'm still working on this, albeit in reduced capacity next to my main work projects. Rather than putting everything into one big changeset and throwing it over the fence, I have decided to try incrementally porting over modules to nanobind.

The biggest roadblock I see might actually be the switch to `nb::ndarray` for the NumPy/Pandas stuff, but I haven't gotten there yet.

**Reply by Tishj** [reply]
I think we can also desugar the ndarray stuff back into py::object and py::handle, it's not used in many places and we don't care that deeply about preserving it as long as it functions the same

**Reply by nicholasjng** [reply]
Yep, but not as easily as just implicitly converting, as `nb::ndarray` is not a subclass of `nb::object` (e.g. things like `nb::module_::import_("numpy").attr("unique")(column)` require an `nb::cast<nb::ndarray>`, also resizing is not yet implemented.)

**Reply by Tishj** [reply]
I'm interested in the nanobind stuff but I'm not sure how much value it adds at the cost of potential incompatibilities.
I would prefer the most maintained one of the two, especially seeing as they provide the exact same API, because I find the API some of the most annoying to deal with.

Especially when you're trying to track reference counts of an object and you're working with a mix of `py::object`, `py::handle`, `py::dict`, `PyObject*`

Just recently I learned the unintuitive reality that `py::reinterpret_borrow<py::object>(handle)` will **increase** the refcount.
Which makes sense from the perspective of "creating an object increases refcount, destroying an object decreases refcount"
But makes absolutely no sense from the perspective of "this handle does not have a reference, because of the type system I *need* this thing to be a py::object, I want to do this without messing with reference counts"

I feel like the API could use a lower level version/control and documentation that explains what the effect of every interaction between objects is through this lower level API.
You could argue the C API is that lower level API but to work with pybind/nanobind we **need** to turn things into handles and objects to work with it, so no

I apologize for the rant
